File: docs/old/clause-and-effect/worksheet-09-searching-a-cyclic-graph.js
let logic = require ("logic-db")

let a = new logic.db_t

a.i (["g", "h"])
a.i (["d", "a"])
a.i (["g", "d"])
a.i (["e", "d"])
a.i (["h", "f"])
a.i (["e", "f"])
a.i (["a", "e"])
a.i (["a", "b"])
a.i (["b", "f"])
a.i (["b", "c"])
a.i (["f", "c"])

// a e d -- in a loop

// we need to avoid searching in a loop

// prepare the lispy list

function cons (x, y) { return { car: x, cdr: y } }

let path = new logic.db_t

path.i (["?x", "?x", "?occurs"])
path.i (["?x", "?y", "?occurs"]) .cond ((the) => {
    let z = new logic.var_t
    return a.o ([the.var.x, z])
        .and (legal.o ([z, the.var.occurs]))
        .and (path.o ([z, the.var.y, cons (z, the.var.occurs)]))
})

let legal = new logic.db_t

legal.i (["?z", null])
legal.i (["?z", cons ("?car", "?cdr")]) .cond ((the) => {
    return legal.o ([the.var.z, the.var.cdr])
        .not_eqv (the.var.z, the.var.car)
})

path.query_log (100) (["f", "f", null])
path.query_log (100) (["a", "c", null])
path.query_log (100) (["g", "e", null])

path.query_log (100) (["g", "?x", null])
path.query_log (100) (["?x", "h", null])

path.query_log (100) (["g", "c", null])
path.query_log (100) (["g", "c", cons ("f", null)])

path.query_log (100) (["a", "?x", cons ("f", cons ("d", null))])
path.query_log (100) (["a", "?x", null])

File: docs/old/clause-and-effect/in-prolog/worksheet-09-searching-a-cyclic-graph.pl
a(g, h).
a(d, a).
a(g, d).
a(e, d).
a(h, f).
a(e, f).
a(a, e).
a(a, b).
a(b, f).
a(b, c).
a(f, c).

%% a e d -- in a loop

%% the following relation
%%   will make query searching in loop
%% path(X, X).
%% path(X, Y) :- a(X, Z), path(Z, Y).

path(X, X, _).
path(X, Y, T) :-
  a(X, Z), legal(Z, T), path(Z, Y, [Z | T]).

legal(_, []).
legal(Z, [H | T]) :- Z \== H, legal(Z, T).

%% path(f, f, []).
%% path(a, c, []).
%% path(g, e, []).
%% path(g, X, []).
%% path(X, h, []).

%% path(g, c, []).
%% path(g, c, [f]).
%% path(a, X, [f, d]).
%% path(a, X, []).
